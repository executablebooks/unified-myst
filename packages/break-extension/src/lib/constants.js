/**
 * The token types for the mystBreak extension.
 */

export const tokenTypes = {
    mystBreak: 'mystBreak',
    mystBreakMarker: 'mystBreakMarker',
}
