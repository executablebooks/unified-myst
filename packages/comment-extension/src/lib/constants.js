/**
 * The token types for the mystComment extension.
 */

export const tokenTypes = {
    mystComment: 'mystComment',
    mystCommentMarker: 'mystCommentMarker',
    mystCommentString: 'mystCommentString',
}
