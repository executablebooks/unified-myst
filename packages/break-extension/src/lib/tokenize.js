/**
 * @typedef {import('micromark-util-types').Extension} Extension
 * @typedef {import('micromark-util-types').Tokenizer} Tokenizer
 * @typedef {import('micromark-util-types').State} State
 */

import { factorySpace } from 'micromark-factory-space'
import { markdownLineEnding, markdownSpace } from 'micromark-util-character'
import { codes } from 'micromark-util-symbol/codes.js'
import { constants } from 'micromark-util-symbol/constants.js'
import { types } from 'micromark-util-symbol/types.js'
import { ok as assert } from 'uvu/assert'
import { tokenTypes } from './constants.js'

/**
 * The micromark extension, to parse the source syntax to events.
 * Events are token enter/exits, which in-turn are references to slices of the source text.
 *
 * @type {Extension}
 */
export const mystBreakMmarkExt = {
    flow: {
        [codes.plusSign]: {
            name: 'mystBreak',
            tokenize: tokenizeMystBreak,
        },
    },
}

const mystBreakMarkerCountMin = 3

/** @type {Tokenizer} */
function tokenizeMystBreak(effects, ok, nok) {
    let size = 0
    return start

    /** @type {State} */
    function start(code) {
        assert(code === codes.plusSign, 'expected `+`')
        effects.enter(tokenTypes.mystBreak)
        return consumeMarker(code)
    }

    /** @type {State} */
    function consumeMarker(code) {
        if (code === codes.plusSign) {
            effects.enter(tokenTypes.mystBreakMarker)
            return sequence(code)
        }

        if (markdownSpace(code)) {
            return factorySpace(effects, consumeMarker, types.whitespace)(code)
        }

        if (size < mystBreakMarkerCountMin) {
            return nok(code)
        }

        if (code === codes.eof || markdownLineEnding(code)) {
            effects.exit(tokenTypes.mystBreak)
            return ok(code)
        }

        // start parsing the content of the comment
        effects.enter(types.chunkString, {
            contentType: constants.contentTypeString,
        })
        return consumeContent(code)
    }

    /** @type {State} */
    function sequence(code) {
        if (code === codes.plusSign) {
            effects.consume(code)
            size++
            return sequence
        }
        effects.exit(tokenTypes.mystBreakMarker)
        return consumeMarker(code)
    }

    /** @type {State} */
    function consumeContent(code) {
        if (code === codes.eof || markdownLineEnding(code)) {
            effects.exit(types.chunkString)
            effects.exit(tokenTypes.mystBreak)
            return ok(code)
        }
        // Consume content of the comment
        effects.consume(code)
        return consumeContent
    }
}
