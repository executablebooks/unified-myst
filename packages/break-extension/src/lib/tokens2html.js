/**
 * @typedef {import('micromark-util-types').HtmlExtension} HtmlExtension
 * @typedef {import('micromark-util-types').Handle} Handle
 */

import { tokenTypes } from './constants.js'

/**
 * The micromark HTML extension, to convert mystBreak tokens directly to HTML.
 * @type {HtmlExtension}
 */
export const mystBreakHtmlExt = {
    enter: { [tokenTypes.mystBreak]: enterMystBreak },
    exit: { [tokenTypes.mystBreak]: exitMystBreak },
}

/**
 * @type {Handle}
 */
function enterMystBreak() {
    this.buffer()
}

/**
 * @type {Handle}
 */
function exitMystBreak() {
    this.resume()
    this.lineEndingIfNeeded()
    this.tag('<hr class="myst-break"/>')
}
