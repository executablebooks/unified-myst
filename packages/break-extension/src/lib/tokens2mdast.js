/**
 * @typedef {import('mdast-util-from-markdown').Extension} FromMarkdownExtension
 * @typedef {import('mdast-util-from-markdown').Handle} FromMarkdownHandle
 *
 * @typedef {any} MystBreakNode
 *   mdast node type for mystBreak
 */

// TODO how to properly define the MystBreakNode with JSDoc?

import { tokenTypes } from './constants.js'

/**
 * The MDAST extension, to convert mystBreak tokens to MDAST nodes.
 * @type {FromMarkdownExtension}
 */
export const mystBreakMdastExt = {
    enter: {
        [tokenTypes.mystBreak]: enterMystBreak,
    },
    exit: {
        [tokenTypes.mystBreak]: exitMystBreak,
    },
}

/**
 * @type {FromMarkdownHandle}
 */
function enterMystBreak(token) {
    this.enter(/** @type {MystBreakNode} */ ({ type: 'mystBreak' }), token)
    this.buffer()
}

/**
 * @type {FromMarkdownHandle}
 */
function exitMystBreak(token) {
    const data = this.resume()
    const node = this.stack[this.stack.length - 1]
    // @ts-ignore
    node.meta = data
    this.exit(token)
}
