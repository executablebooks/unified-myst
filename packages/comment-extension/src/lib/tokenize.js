/**
 * @typedef {import('micromark-util-types').Construct} Construct
 * @typedef {import('micromark-util-types').Extension} Extension
 * @typedef {import('micromark-util-types').Tokenizer} Tokenizer
 * @typedef {import('micromark-util-types').State} State
 */

import { factorySpace } from 'micromark-factory-space'
import { markdownLineEnding } from 'micromark-util-character'
import { codes } from 'micromark-util-symbol/codes.js'
import { constants } from 'micromark-util-symbol/constants.js'
import { types } from 'micromark-util-symbol/types.js'
import { ok as assert } from 'uvu/assert'
import { tokenTypes } from './constants.js'

/**
 * The micromark extension, to parse the source syntax to events.
 * Events are token enter/exits, which in-turn are references to slices of the source text.
 *
 * @type {Extension}
 */
export const mystCommentMmarkExt = {
    flow: {
        [codes.percentSign]: {
            name: 'mystComment',
            tokenize: tokenizeMystComment,
        },
    },
}

/** @type {Construct} */
const nextLine = { tokenize: tokenizeNextLine, partial: true }

/** @type {Tokenizer} */
function tokenizeMystComment(effects, ok) {
    return start

    /** @type {State} */
    function start(code) {
        assert(code === codes.percentSign, 'expected `%`')
        effects.enter(tokenTypes.mystComment)
        effects.enter(tokenTypes.mystCommentMarker)
        effects.consume(code)
        effects.exit(tokenTypes.mystCommentMarker)
        return afterMarker
    }

    /** @type {State} */
    function afterMarker(code) {
        if (code === codes.eof) {
            return finalise(code)
        }

        if (markdownLineEnding(code)) {
            return effects.attempt(nextLine, afterMarker, finalise)
        }

        // start parsing the content of the comment
        effects.enter(types.chunkString, {
            contentType: constants.contentTypeString,
        })
        return insideContent(code)
    }

    /** @type {State} */
    function insideContent(code) {
        if (code === codes.eof) {
            effects.exit(types.chunkString)
            return finalise(code)
        }

        if (markdownLineEnding(code)) {
            effects.exit(types.chunkString)
            return effects.attempt(nextLine, afterMarker, finalise)
        }

        // Consume content of the comment
        effects.consume(code)
        return insideContent
    }

    /** @type {State} */
    function finalise(code) {
        effects.exit(tokenTypes.mystComment)
        return ok(code)
    }
}

/** @type {Tokenizer} */
function tokenizeNextLine(effects, ok, nok) {
    return lineEnd

    /** @type {State} */
    function lineEnd(code) {
        // Consume the previous line ending, then allow for up to three spaces
        effects.enter(types.lineEnding)
        effects.consume(code)
        effects.exit(types.lineEnding)
        return factorySpace(
            effects,
            afterPrefix,
            types.linePrefix,
            constants.tabSize - 1
        )
    }

    /** @type {State} */
    function afterPrefix(code) {
        if (code !== codes.percentSign) {
            return nok(code)
        }
        effects.enter(tokenTypes.mystCommentMarker)
        effects.consume(code)
        effects.exit(tokenTypes.mystCommentMarker)
        return ok(code)
    }
}
