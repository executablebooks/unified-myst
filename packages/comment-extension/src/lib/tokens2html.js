/**
 * @typedef {import('micromark-util-types').HtmlExtension} HtmlExtension
 * @typedef {import('micromark-util-types').Handle} Handle
 */

import { tokenTypes } from './constants.js'

/**
 * The micromark HTML extension, to convert mystComment tokens directly to HTML.
 * @type {HtmlExtension}
 */
export const mystCommentHtmlExt = {
    enter: { [tokenTypes.mystComment]: enterMystComment },
    exit: { [tokenTypes.mystComment]: exitMystComment },
}

/**
 * @type {Handle}
 */
function enterMystComment() {
    this.buffer()
}

/**
 * @type {Handle}
 */
function exitMystComment() {
    const content = this.encode(this.resume())
    this.tag('<!-- ')
    this.raw(content.trim())
    this.tag(' -->')
}
