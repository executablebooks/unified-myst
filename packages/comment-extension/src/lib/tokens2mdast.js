/**
 * @typedef {import('mdast-util-from-markdown').Extension} FromMarkdownExtension
 * @typedef {import('mdast-util-from-markdown').Handle} FromMarkdownHandle
 *
 * @typedef {any} MystCommentNode
 *   mdast node type for mystComment
 */

// TODO how to properly define the MystCommentNode with JSDoc?

import { tokenTypes } from './constants.js'

/**
 * The MDAST extension, to convert mystComment tokens to MDAST nodes.
 * @type {FromMarkdownExtension}
 */
export const mystCommentMdastExt = {
    enter: {
        [tokenTypes.mystComment]: enterMystComment,
    },
    exit: {
        [tokenTypes.mystComment]: exitMystComment,
    },
}

/**
 * @type {FromMarkdownHandle}
 */
function enterMystComment(token) {
    this.enter(/** @type {MystCommentNode} */ ({ type: 'comment' }), token)
    this.buffer()
}

/**
 * @type {FromMarkdownHandle}
 */
function exitMystComment(token) {
    const data = this.resume()
    const node = this.stack[this.stack.length - 1]
    // @ts-ignore
    node.value = data
    this.exit(token)
}
