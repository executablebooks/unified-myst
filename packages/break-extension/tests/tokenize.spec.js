import { parse } from 'micromark/lib/parse'
import { postprocess } from 'micromark/lib/postprocess'
import { preprocess } from 'micromark/lib/preprocess'

import { mystBreakMmarkExt } from '../src/index.js'

const fixtures = [
    '+',
    '+++',
    '+++\n',
    '+++{}',
    '+++{}\n',
    '+++ {"key": "value"}',
]

describe('Convert source to tokens', () => {
    test.each(fixtures)(`%j`, (content) => {
        const options = { extensions: [mystBreakMmarkExt] }
        const events = postprocess(
            parse(options).document().write(preprocess()(content, 'utf8', true))
        )
        const tokens = events.map((event) => {
            return [
                event[0],
                event[1].type,
                `${event[1].start.line}.${event[1].start.column}`,
                `${event[1].end.line}.${event[1].end.column}`,
            ]
        })
        expect(tokens).toMatchSnapshot()
    })
})
