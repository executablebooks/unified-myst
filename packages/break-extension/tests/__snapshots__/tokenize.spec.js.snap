// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert source to tokens "+" 1`] = `
Array [
  Array [
    "enter",
    "listUnordered",
    "1.1",
    "1.2",
  ],
  Array [
    "enter",
    "listItemPrefix",
    "1.1",
    "1.2",
  ],
  Array [
    "enter",
    "listItemMarker",
    "1.1",
    "1.2",
  ],
  Array [
    "exit",
    "listItemMarker",
    "1.1",
    "1.2",
  ],
  Array [
    "exit",
    "listItemPrefix",
    "1.1",
    "1.2",
  ],
  Array [
    "exit",
    "listUnordered",
    "1.1",
    "1.2",
  ],
]
`;

exports[`Convert source to tokens "+++ {\\"key\\": \\"value\\"}" 1`] = `
Array [
  Array [
    "enter",
    "mystBreak",
    "1.1",
    "1.21",
  ],
  Array [
    "enter",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "whitespace",
    "1.4",
    "1.5",
  ],
  Array [
    "exit",
    "whitespace",
    "1.4",
    "1.5",
  ],
  Array [
    "enter",
    "data",
    "1.5",
    "1.21",
  ],
  Array [
    "exit",
    "data",
    "1.5",
    "1.21",
  ],
  Array [
    "exit",
    "mystBreak",
    "1.1",
    "1.21",
  ],
]
`;

exports[`Convert source to tokens "+++" 1`] = `
Array [
  Array [
    "enter",
    "mystBreak",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreak",
    "1.1",
    "1.4",
  ],
]
`;

exports[`Convert source to tokens "+++\\n" 1`] = `
Array [
  Array [
    "enter",
    "mystBreak",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreak",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "lineEnding",
    "1.4",
    "2.1",
  ],
  Array [
    "exit",
    "lineEnding",
    "1.4",
    "2.1",
  ],
]
`;

exports[`Convert source to tokens "+++{}" 1`] = `
Array [
  Array [
    "enter",
    "mystBreak",
    "1.1",
    "1.6",
  ],
  Array [
    "enter",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "data",
    "1.4",
    "1.6",
  ],
  Array [
    "exit",
    "data",
    "1.4",
    "1.6",
  ],
  Array [
    "exit",
    "mystBreak",
    "1.1",
    "1.6",
  ],
]
`;

exports[`Convert source to tokens "+++{}\\n" 1`] = `
Array [
  Array [
    "enter",
    "mystBreak",
    "1.1",
    "1.6",
  ],
  Array [
    "enter",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "mystBreakMarker",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "data",
    "1.4",
    "1.6",
  ],
  Array [
    "exit",
    "data",
    "1.4",
    "1.6",
  ],
  Array [
    "exit",
    "mystBreak",
    "1.1",
    "1.6",
  ],
  Array [
    "enter",
    "lineEnding",
    "1.6",
    "2.1",
  ],
  Array [
    "exit",
    "lineEnding",
    "1.6",
    "2.1",
  ],
]
`;
